paymentsheet/src/androidTest/java/com/stripe/android/paymentsheet/GooglePayTest.kt

```diff
@@ -38,52 +60,189 @@ internal class GooglePayTest {
 
     private val composeTestRule = testRules.compose
 
+    @Before
+    fun setup() {
+        Intents.init()
+    }
+
     @After
     fun teardown() {
         GooglePayRepository.resetFactory()
+        Intents.release()
     }
 
     @Test
-    fun googlePayIsShownWhenFullyEnabled() = runGooglePayTest(
+    fun googlePayIsShownWhenFullyEnabled() = runGooglePayAvailabilityTest(
         isGooglePayReady = true,
         isGooglePayEnabledInElementsSession = true,
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertExists()
-    }
+        shouldHaveGooglePayOption = true,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenNotReady() = runGooglePayTest(
+    fun googlePayIsNotShownWhenNotReady() = runGooglePayAvailabilityTest(
         isGooglePayReady = false,
         isGooglePayEnabledInElementsSession = true,
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertDoesNotExist()
-    }
+        shouldHaveGooglePayOption = false,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenNotEnabledInElementsSession() = runGooglePayTest(
+    fun googlePayIsNotShownWhenNotEnabledInElementsSession() = runGooglePayAvailabilityTest(
         isGooglePayReady = true,
         isGooglePayEnabledInElementsSession = false,
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertDoesNotExist()
-    }
+        shouldHaveGooglePayOption = false,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenGooglePayConfigIsNotProvided() = runGooglePayTest(
+    fun googlePayIsNotShownWhenGooglePayConfigIsNotProvided() = runGooglePayAvailabilityTest(
         isGooglePayReady = true,
         isGooglePayEnabledInElementsSession = true,
         hasGooglePayConfig = false,
+        shouldHaveGooglePayOption = false,
+    )
+
+    @Test
+    fun googlePaySucceeds() {
+        var resultCallbackCalled = false
+
+        runGooglePayFlowTest(
+            paymentResultCallback = {
+                resultCallbackCalled = true
+
+                assertCompleted(it)
+            }
+        ) { scenario ->
+            val paymentMethod = PaymentMethodFactory.card()
+
+            intendingGooglePayToBeLaunched(GooglePayPaymentMethodLauncher.Result.Completed(paymentMethod))
+            intendingPaymentConfirmationToBeLaunched(
+                InternalPaymentResult.Completed(PaymentIntentFactory.create(paymentMethod))
+            )
+
+            scenario.confirm()
+
+            intendedGooglePayToBeLaunched()
+            intendedPaymentConfirmationToBeLaunched()
+
+            composeTestRule.waitUntil(UI_TIMEOUT) {
```

Comment:
Seems like it would be better to use a countdown latch here.

Reply:
I actually need this in order to have the full screen animation to run completely